### PR TITLE
Create native feature runner that can explicitly run a scenario from a test

### PR DESCRIPTION
**Pull request by jacdevos**
I created a direct, explicit native test runner for scenarios, based on my observation in Issue 
https://github.com/net-a-porter-mobile/XCTest-Gherkin/issues/49

It allows you to create normal XCTests, and run native feature files explicitly, as in:
func testNativeExampleWithScenarioRunner() {
  NativeRunner.runScenario(featureFile: "NativeFeatures/native_example.feature", scenario: "This is a basic happy path example", testCase: self)
}

This is just V0.001 so please give suggestions if the idea has wings.

### CODE REVIEW

File: Pod/Native/NativeRunner.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+//Gives us the ability to run features or scenarios directly by specifying file and name
+open class NativeRunner {
+    
+    public class func runScenario(featureFile: String, scenario: String?, testCase: XCTestCase) {
+        testCase.state.loadAllStepsIfNeeded()
+        let path = (Bundle(for: type(of: testCase)).resourceURL?.appendingPathComponent(featureFile))!
+        let features = loadFeatures(path: path)
+        
+        for feature in features {
+            let scenarios = feature.scenarios.filter({ scenario == nil || $0.scenarioDescription.hasPrefix(scenario!)});
```
Comment (kerrmarin) [inline Pod/Native/NativeRunner.swift]: What do you think of making this not inline and removing the semicolon?

File: Pod/Native/NativeRunner.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            if scenarios.count < 1{
```
Comment (kerrmarin) [inline Pod/Native/NativeRunner.swift]: Nitpicky, but you're missing a space between `1` and `{`

File: Pod/Native/NativeRunner.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+    
+    
+    public class func runFeature(featureFile: String, testCase: XCTestCase) {
+        NativeRunner.runScenario(featureFile: featureFile, scenario: nil, testCase: testCase)
+    }
+
+    
+    private class func loadFeatures(path : URL) -> [NativeFeature] {
```
Comment (kerrmarin) [inline Pod/Native/NativeRunner.swift]: nitpicky, but can we remove the space between `path` and `:`

File: Pod/Native/NativeRunner.swift [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        let features = loadFeatures(path: path)
+        
+        for feature in features {
+            let scenarios = feature.scenarios.filter({ scenario == nil || $0.scenarioDescription.hasPrefix(scenario!)});
+            if scenarios.count < 1{
+                XCTFail("No scenario found with name: \(scenario)")
+            }
+            
+            for scenario in scenarios {
+                let allScenarioStepsDefined = scenario.stepDescriptions.map(testCase.state.matchingGherkinStepExpressionFound).reduce(true) { $0 && $1 }
```
Comment (kerrmarin) [inline Pod/Native/NativeRunner.swift]: Nice

**Comment by smaljaar** [comment]
@jacdevos could you plz update the PR so this can be merged? it's been open for a while

**Comment by nap-sam-dean** [comment]
@smaljaar If you're interested in this getting into master I don't mind merging it and fixing it myself?

**Comment by jacdevos** [comment]
@smaljaar my bad, I'll make some time for it this week, it has been quite useful on our projects.

**Comment by jacdevos** [comment]
@smaljaar I've pushed the style you suggested. I also have a few tests in ExampleNativeRunnerTest.

I think we are ready to go. I have a few additions to this that came up in our project, but will add it in a new PR.

**Comment by kerrmarin** [comment]
Thanks @jacdevos I'll investigate the test failures later today

**Comment by kerrmarin** [comment]
Looks like it was fixed by running the tests again :)

File: Pod/Native/NativeRunner.swift [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        for feature in features {
+            let scenarios = feature.scenarios.filter({
+                scenario == nil || $0.scenarioDescription.hasPrefix(scenario!)
+            })
+            
+            if scenarios.count < 1 {
+                XCTFail("No scenario found with name: \(scenario)")
+            }
+            
+            for scenario in scenarios {
```
Comment (smaljaar) [inline Pod/Native/NativeRunner.swift]: This for loop is duplicate code copied from the NativeTestCase method featureScenarioTest(). I extracted the code from featureScenarioTest() into this auxiliary method in NativeTestCase so we avoid having the same code in two places:

      class func perform(scenario: NativeScenario, from feature: NativeFeature, in testCase: XCTestCase) {
        let allScenarioStepsDefined = scenario.stepDescriptions.map(testCase.state.matchingGherkinStepExpressionFound).reduce(true) { $0 && $1 }
        var allFeatureBackgroundStepsDefined = true
        
        if let defined = feature.background?.stepDescriptions.map(testCase.state.matchingGherkinStepExpressionFound).reduce(true, { $0 && $1 }) {
            allFeatureBackgroundStepsDefined = defined
        }
        
        guard allScenarioStepsDefined && allFeatureBackgroundStepsDefined else {
            XCTFail("Some step definitions not found for the scenario: \(scenario.scenarioDescription)")
            return
        }
        
        if let background = feature.background {
            background.stepDescriptions.forEach(testCase.performStep)
        }
        scenario.stepDescriptions.forEach(testCase.performStep)
    }
    
You can now replace the body of this for loop with the line:

        NativeRunner.runScenario(featureFile: featureFile, scenario: nil, testCase: testCase)

In NativeTestCase you can call this method like:

        NativeTestCase.perform(scenario: scenario, from: feature, in: self)


If you like the suggestion, please use it 🙂

**Comment by smaljaar** [comment]
@jacdevos I studied the NativeRunner class and I really like how you solved the problem of disappearing test cases in the Xcode Test Navigator 👍

File: Pod/Native/NativeRunner.swift (hunk 5 above)
Comment (jacdevos) [inline Pod/Native/NativeRunner.swift]: Good suggestion @smaljaar - maybe create another PR for this, I see change has been merged.

**Comment by jacdevos** [comment]
@smaljaar glad you like it - it's been working quite well for us. There are a few enhancements that would help, which we can work on if we see the value:
• seems like we can remove some boilerplate 
• a generator would be really nice: feed it the feature file and spit out a XCTest class with one test per scenario and a steps file with blank steps

**Comment by nap-sam-dean** [comment]
@jacdevos Nice ideas - I'll make the generator one into an issue - I think I've heard someone else talk about this too, let's see if there's interest. And if you can see ways to reduce boilerplate - that would be fantastic!

**Pull request merged**